File: Assets/Plugins/Android/src/com/playphone/multinet/unity/MNVShopProviderUnity.java
package com.playphone.multinet.unity;

import com.playphone.multinet.*;
import com.unity3d.player.UnityPlayer;
import com.playphone.multinet.providers.*;

public class MNVShopProviderUnity {
    public static void shutdown() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVShopProvider().shutdown();
            }
        });
    }

    public static String getVShopPackList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getVShopProvider().getVShopPackList());
    }

    public static String getVShopCategoryList() {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getVShopProvider().getVShopCategoryList());
    }

    public static String findVShopPackById(final int packId) {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getVShopProvider().findVShopPackById(packId));
    }

    public static String findVShopCategoryById(final int categoryId) {
        MNUnity.MARK();

        return MNUnity.serializer.serialize(MNDirect.getVShopProvider().findVShopCategoryById(categoryId));
    }

    public static boolean isVShopInfoNeedUpdate() {
        MNUnity.MARK();

        return MNDirect.getVShopProvider().isVShopInfoNeedUpdate();
    }

    public static void doVShopInfoUpdate() {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVShopProvider().doVShopInfoUpdate();
            }
        });
    }

    public static String getVShopPackImageURL(final int packId) {
        MNUnity.MARK();

        return MNDirect.getVShopProvider().getVShopPackImageURL(packId);
    }

    public static boolean isVShopReady() {
        MNUnity.MARK();

        return MNDirect.getVShopProvider().isVShopReady();
    }

    public static void execCheckoutVShopPacks(final String packIdArray, final String packCountArray, final long clientTransactionId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVShopProvider().execCheckoutVShopPacks(MNUnity.serializer.deserialize(packIdArray,int[].class), MNUnity.serializer.deserialize(packCountArray,int[].class), clientTransactionId);
            }
        });
    }

    public static void procCheckoutVShopPacksSilent(final String packIdArray, final String packCountArray, final long clientTransactionId) {
        MNUnity.MARK();

        UnityPlayer.currentActivity.runOnUiThread(new Runnable()
        {
            public void run()
            {
                MNDirect.getVShopProvider().procCheckoutVShopPacksSilent(MNUnity.serializer.deserialize(packIdArray,int[].class), MNUnity.serializer.deserialize(packCountArray,int[].class), clientTransactionId);
            }
        });
    }

    protected static class MNVShopProviderEventHandler implements com.playphone.multinet.providers.MNVShopProvider.IEventHandler {

        @Override
        public void onVShopInfoUpdated() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onVShopInfoUpdated");
                }
            });
        }

        @Override
        public void showDashboard() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_showDashboard");
                }
            });
        }

        @Override
        public void hideDashboard() {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_hideDashboard");
                }
            });
        }

        @Override
        public void onCheckoutVShopPackSuccess(final MNVShopProvider.IEventHandler.CheckoutVShopPackSuccessInfo result) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onCheckoutVShopPackSuccess",result);
                }
            });
        }

        @Override
        public void onCheckoutVShopPackFail(final MNVShopProvider.IEventHandler.CheckoutVShopPackFailInfo result) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onCheckoutVShopPackFail",result);
                }
            });
        }

        @Override
        public void onVShopReadyStatusChanged(final boolean isVShopReady) {
            MNUnity.MARK();

            UnityPlayer.currentActivity.runOnUiThread(new Runnable()
            {
                public void run()
                {
                  MNUnity.callUnityFunction("MNUM_onVShopReadyStatusChanged",isVShopReady);
                }
            });
        }
    }

    private static MNVShopProviderEventHandler registeredEventHandler = null;
    public static synchronized boolean registerEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getVShopProvider() == null) {
            MNUnity.ELog("MNVShopProvider is not ready. Use MNDirect\'s sessionReady event for adding your delegates.");
            return false;
        }

        if (registeredEventHandler == null) {
            registeredEventHandler = new MNVShopProviderUnity.MNVShopProviderEventHandler();
            MNDirect.getVShopProvider().addEventHandler(registeredEventHandler);
        }

        return true;
    }
    
    public static synchronized boolean unregisterEventHandler() {
        MNUnity.MARK();

        if (MNDirect.getVShopProvider() == null) {
            return false;
        }

        if (registeredEventHandler != null) {
            MNDirect.getVShopProvider().removeEventHandler(registeredEventHandler);
            registeredEventHandler = null;
        }

        return true;
    }
}

File: Assets/Plugins/iOS/PlayPhoneMultiNet/unity/providers/MNVShopProviderUnity.m
//
//  MNVShopProviderUnity.m
//  PlayPhone SDK for Unity
//
//  Copyright (c) 2012 PlayPhone Inc. All rights reserved.
//

#import "MNDirect.h"
#import "MNVShopProvider.h"

#import "MNVShopProviderUnity.h"

@interface MNVShopProviderUnity : NSObject <MNVShopProviderDelegate>
+ (id)shared;
- (void)onVShopInfoUpdated;
- (void)showDashboard;
- (void)hideDashboard;
- (void)onCheckoutVShopPackSuccess:(MNVShopProviderCheckoutVShopPackSuccessInfo*)result;
- (void)onCheckoutVShopPackFail:(MNVShopProviderCheckoutVShopPackFailInfo*)result;
- (void)onVShopReadyStatusChanged:(bool)isVShopReady;
@end

static MNVShopProviderUnity *MNVShopProviderUnityInstance = nil;

@implementation MNVShopProviderUnity

+(id)shared {
    if (MNVShopProviderUnityInstance == nil) {
        MNVShopProviderUnityInstance = [[MNVShopProviderUnity alloc]init];
    }
    
    return MNVShopProviderUnityInstance;
}

- (void)onVShopInfoUpdated {
    [MNUnity callUnityFunction:@"MNUM_onVShopInfoUpdated" withParams:nil];
}
- (void)showDashboard {
    [MNUnity callUnityFunction:@"MNUM_showDashboard" withParams:nil];
}
- (void)hideDashboard {
    [MNUnity callUnityFunction:@"MNUM_hideDashboard" withParams:nil];
}
- (void)onCheckoutVShopPackSuccess:(MNVShopProviderCheckoutVShopPackSuccessInfo*)result {
    [MNUnity callUnityFunction:@"MNUM_onCheckoutVShopPackSuccess" withParams:(result == nil ? [NSNull null] : result), nil];
}
- (void)onCheckoutVShopPackFail:(MNVShopProviderCheckoutVShopPackFailInfo*)result {
    [MNUnity callUnityFunction:@"MNUM_onCheckoutVShopPackFail" withParams:(result == nil ? [NSNull null] : result), nil];
}
- (void)onVShopReadyStatusChanged:(bool)isVShopReady {
    [MNUnity callUnityFunction:@"MNUM_onVShopReadyStatusChanged" withParams:[NSNumber numberWithBool:isVShopReady], nil];
}
@end

EXTERN_C
const char* _MNVShopProvider_GetVShopPackList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect vShopProvider] getVShopPackList]] UTF8String]);
}

EXTERN_C
const char* _MNVShopProvider_GetVShopCategoryList() {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect vShopProvider] getVShopCategoryList]] UTF8String]);
}

EXTERN_C
const char* _MNVShopProvider_FindVShopPackById(int packId) {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect vShopProvider] findVShopPackById:packId]] UTF8String]);
}

EXTERN_C
const char* _MNVShopProvider_FindVShopCategoryById(int categoryId) {
    MARK;

    return MakeStringCopy([[[MNUnity serializer]serialize:[[MNDirect vShopProvider] findVShopCategoryById:categoryId]] UTF8String]);
}

EXTERN_C
bool _MNVShopProvider_IsVShopInfoNeedUpdate() {
    MARK;

    return [[MNDirect vShopProvider] isVShopInfoNeedUpdate];
}

EXTERN_C
void _MNVShopProvider_DoVShopInfoUpdate() {
    MARK;

    [[MNDirect vShopProvider] doVShopInfoUpdate];
}

EXTERN_C
const char* _MNVShopProvider_GetVShopPackImageURL(int packId) {
    MARK;

    NSURL *resultUrl = [[MNDirect vShopProvider] getVShopPackImageURL:packId];

    if (resultUrl == nil) {
        return NULL;
    }

    return MakeStringCopy([[resultUrl absoluteString] UTF8String]);
}

EXTERN_C
bool _MNVShopProvider_IsVShopReady() {
    MARK;

    return [[MNDirect vShopProvider] isVShopReady];
}

EXTERN_C
void _MNVShopProvider_ExecCheckoutVShopPacks(const char* packIdArray, const char* packCountArray, long long clientTransactionId) {
    MARK;

    [[MNDirect vShopProvider] execCheckoutVShopPacks:[[MNUnity serializer] deserializeArray:[NSNumber class] fromJson:[NSString stringWithUTF8String:packIdArray]] packCount:[[MNUnity serializer] deserializeArray:[NSNumber class] fromJson:[NSString stringWithUTF8String:packCountArray]] clientTransactionId:clientTransactionId];
}

EXTERN_C
void _MNVShopProvider_ProcCheckoutVShopPacksSilent(const char* packIdArray, const char* packCountArray, long long clientTransactionId) {
    MARK;

    [[MNDirect vShopProvider] procCheckoutVShopPacksSilent:[[MNUnity serializer] deserializeArray:[NSNumber class] fromJson:[NSString stringWithUTF8String:packIdArray]] packCount:[[MNUnity serializer] deserializeArray:[NSNumber class] fromJson:[NSString stringWithUTF8String:packCountArray]] clientTransactionId:clientTransactionId];
}

EXTERN_C
bool _MNVShopProvider_RegisterEventHandler(){
    MARK;

    if ([MNDirect vShopProvider] == nil) {
        return false;
    }

    [[MNDirect vShopProvider] addDelegate:[MNVShopProviderUnity shared]];

    return true;
}

EXTERN_C
bool _MNVShopProvider_UnregisterEventHandler(){
    MARK;

    if ([MNDirect vShopProvider] == nil) {
        return false;
    }

    [[MNDirect vShopProvider] removeDelegate:[MNVShopProviderUnity shared]];

    return true;
}


